### PR TITLE
perf(flags): add pagination to flags list

### LINKED ISSUE #3787: flarum/flags should use pagination
### Current Behavior

Currently in one of the forums I manage I have about 450 flagged posts that are somewhat stagnant, the thing is that when you click on the dropdown it takes some time to display the list (even if I remember correctly after 150 you can start to notice a slowdown), I think it would be good to introduce pagination to avoid this, similar to the one used in the notifications dropdown.

### Steps to Reproduce

1. Flag several hundred posts
2. Open the flags dropdown

### Screenshots

_No response_

### Environment

- Flarum version: 1.7.2
- PHP version: 8.2.4

### Possible Solution

Make use pagination to improve performance and avoid lots of requests.

### PR DESCRIPTION
**Pull request by OrdinaryJellyfish**
**Fixes #3787**

**Changes proposed in this pull request:**

* Adds pagination to flags list

**Reviewers should focus on:**
* If the pagination works properly
* If flags are ordered properly

**Screenshot**

**QA**

**Necessity**

- [ ] Has the problem that is being solved here been clearly explained?
- [ ] If applicable, have various options for solving this problem been considered?
- [ ] For core PRs, does this need to be in core, or could it be in an extension?
- [ ] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).
- [ ] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

### CODE REVIEW

File: extensions/flags/js/src/forum/components/FlagList.tsx [hunk 1]
```diff
@@ -16,49 +18,53 @@ export interface IFlagListAttrs extends ComponentAttrs {
...
         className="FlagList"
         title={app.translator.trans('flarum-flags.forum.flagged_posts.title')}
-        hasItems={flags.length}
-        loading={this.state.loading}
+        hasItems={state.hasItems()}
+        loading={state.isLoading()}
         emptyText={app.translator.trans('flarum-flags.forum.flagged_posts.empty_text')}
+        loadMore={() => state.hasNext() && !state.isLoadingNext() && state.loadNext()}
       >
-        <ul className="HeaderListGroup-content">
```
Comment (SychO9) [inline extensions/flags/js/src/forum/components/FlagList.tsx:33]: any reason why the `ul` with `li` items is gone now?

File: extensions/flags/src/Api/Controller/ListFlagsController.php [hunk 2]
```diff
@@ -28,26 +29,57 @@ class ListFlagsController extends AbstractListController
...
+            ->groupBy('post_id')
+            ->orderBy('created_at', 'DESC')
+            ->skip($offset)
+            ->take($limit + 1);
+
+        $flags = Flag::whereVisibleTo($actor)
+            ->select('flags.*')
+            ->joinSub($primaries, 'p', 'flags.id', '=', 'p.id')
+            ->latest()
+            ->get();
```
Comment (SychO9) [inline extensions/flags/src/Api/Controller/ListFlagsController.php]: I don't understand the two queries, what does it do and why not a simple paginated query?

**Review by SychO9 — COMMENTED** [review]
Thanks for the PR! as a side not, you usually want to avoid committing anything in dist or dist-typings

File: extensions/flags/js/src/forum/components/FlagList.tsx (hunk 1 above)
Comment (OrdinaryJellyfish) [reply extensions/flags/js/src/forum/components/FlagList.tsx:33]: That was an accident! I will add it back.

File: extensions/flags/src/Api/Controller/ListFlagsController.php (hunk 2 above)
Comment (OrdinaryJellyfish) [reply extensions/flags/src/Api/Controller/ListFlagsController.php]: I just copied how it was done from `NotificationRepository`. When I tried a single query it wouldn't load more past the first 20.

Comment (SychO9) [reply extensions/flags/src/Api/Controller/ListFlagsController.php]: Oh I see. Yea this querying logic is unique to notifications, so what you want to do is recover the old simple flag query and in addition apply a ->limit($limit + 1)->offset($offset) and it should be good to go

Comment (OrdinaryJellyfish) [reply extensions/flags/src/Api/Controller/ListFlagsController.php]: Alrighty!

**Review by SychO9 — APPROVED** [review]
Looks good to me!

**Pull request merged**